frameworks/C++/ffead-cpp/benchmark_config.json

```diff
@@ -1,95 +1,311 @@
-      "plaintext_url": "/te-benchmark/plaintext",
-      "db_url": "/te-benchmark/db",
-      "query_url": "/te-benchmark/queries?queries=",
-      "fortune_url": "/te-benchmark/fortunes",
-      "update_url": "/te-benchmark/updates?queries=",
-      "port": 8080,
-      "approach": "Realistic",
-      "classification": "Fullstack",
-      "database": "mongodb",
-      "framework": "ffead-cpp",
-      "language": "C++",
-      "orm": "Full",
-      "platform": "ffead-cpp",
-      "webserver": "ffead-cpp",
-      "os": "Linux",
-      "database_os": "Linux",
-      "display_name": "ffead-cpp-mongo",
-      "notes": "",
-      "versus": ""
-    },
-	"postgresql": {
-      "json_url": "/te-benchmark/json",
-      "plaintext_url": "/te-benchmark/plaintext",
-      "db_url": "/te-benchmark/db",
-      "query_url": "/te-benchmark/queries?queries=",
-      "fortune_url": "/te-benchmark/fortunes",
-      "update_url": "/te-benchmark/updates?queries=",
-      "port": 8080,
-      "approach": "Realistic",
-      "classification": "Fullstack",
-      "database": "postgres",
-      "framework": "ffead-cpp",
-      "language": "C++",
-      "orm": "Full",
-      "platform": "ffead-cpp",
-      "webserver": "ffead-cpp",
-      "os": "Linux",
-      "database_os": "Linux",
-      "display_name": "ffead-cpp-postgresql",
-      "notes": "",
-      "versus": ""
-    },
-    "apache-mongo": {
-      "setup_file": "setup-apache-mongo",
-      "json_url": "/te-benchmark/json",
-      "plaintext_url": "/te-benchmark/plaintext",
-      "db_url": "/te-benchmark/db",
-      "query_url": "/te-benchmark/queries?queries=",
-      "fortune_url": "/te-benchmark/fortunes",
-      "update_url": "/te-benchmark/updates?queries=",
-      "port": 8080,
-      "approach": "Realistic",
-      "classification": "Fullstack",
-      "database": "mongodb",
-      "framework": "ffead-cpp",
-      "language": "C++",
-      "orm": "Full",
-      "platform": "ffead-cpp",
-      "webserver": "apache",
-      "os": "Linux",
-      "database_os": "Linux",
-      "display_name": "ffead-cpp-apache-mongo",
-      "notes": "",
-      "versus": ""
-    },
-	"apache-postgresql": {
-      "setup_file": "setup-apache-postgresql",
-      "json_url": "/te-benchmark/json",
-      "plaintext_url": "/te-benchmark/plaintext",
-      "db_url": "/te-benchmark/db",
-      "query_url": "/te-benchmark/queries?queries=",
-      "fortune_url": "/te-benchmark/fortunes",
-      "update_url": "/te-benchmark/updates?queries=",
-      "port": 8080,
-      "approach": "Realistic",
-      "classification": "Fullstack",
-      "database": "postgres",
-      "framework": "ffead-cpp",
-      "language": "C++",
-      "orm": "Full",
-      "platform": "ffead-cpp",
-      "webserver": "apache",
-      "os": "Linux",
-      "database_os": "Linux",
-      "display_name": "ffead-cpp-apache-postgresql",
-      "notes": "",
-      "versus": ""
-    }
-  }]
+	"framework": "ffead-cpp",
+	"tests": [{
```

Comment:
It looks like the travis build for ffead hit its 50 minute timeout.  We could work around this by splitting ffead into multiple jobs, but we don't want to do that.  50 minutes is more time than we want to spend building or testing any single framework.

ffead needs to fit within that 50 minute window somehow.  Here are a few ideas for how you could make that happen:
* Consider combining the variants that only implement the cached query test type into the other variants.  For example, can `redis` be combined with `default`?
* Check if you can make better use of docker's build cache by reorganizing the dockerfiles to share larger common prefixes.  For example, if two dockerfiles differ only in one line, try moving that line as far down in both dockerfiles as possible.
* Consider simply removing some of these variants.  For example, consider removing all of the `apache` variants or all of the `nginx` variants, and consider only targeting one database rather than all three, and either redis or memcached but not both.  As the framework's maintainer, you're in a better position than most people to decide which of these variants is best, and readers who find out about your framework by way of TFB may appreciate you being decisive -- "run my framework like this" -- rather than being shown several options.  Consolidating/removing options tends to make these test implementations much less complicated, which makes the frameworks look less imposing and easier to use.

Reply:
Docker cache works for a specific dockerfile and not across multiple dockerfiles. To resolve this issue I have added support for a single docker file with multiple command variations, these commands are defined in the specific test benchmark config files along with the base image name to use. The idea is to build a whole/complete base image for the framework with all bells/whistles and then based on the entrypoint/command specified, change the framework behavior accordingly. Also I would not want to remove any of these tests from the benchmark as the whole point of having this project is to provide various combinations of frameworks/servers/databases to identify better performing ones. I also have plans to add a lighthttpd module down the line, I want ffead-cpp to be the one that supports any server backend, any database any cache, my idea with ffead-cpp is to make it super simple to develop with C++ without worrying too much about the internals. So with this approach I can sneak in more combinations of the framework in the future as well. Do let me know what you think of this approach?

Reply:
If you look at the python code that builds/runs a docker image, you will see that every test flavor is named and this same name is used as the container name, so in short a Dockerfile currently in the framework is equal to a named container. The Docker cache works on the Dockerfile and hence this does not work for multiple dockerfiles correctly. In order for Docker cache to work correctly the child container should be spawned from the same base image (same named container). Btw I already tried doing what you suggested and it did not work out, what I did in short was have 13 different Dockerfiles all similar except their last lines (the CMD). So yes i have tried your suggested approach and then I zeroed in on the this change that you see here. This approach as you said is easier as it entertains the framework maintainers and authors to keep the number of artifacts to a minimum and yes less duplication. The spearate Dockerfile approach takes almost 18*13 minutes of time to run which is way higher that your stipulated 50 min travis runtime.

Reply:
> Btw I already tried doing what you suggested and it did not work out, what I did in short was have 13 different Dockerfiles all similar except their last lines (the CMD). So yes i have tried your suggested approach and then I zeroed in on the this change that you see here.

What you're claiming here conflicts with what I observe locally and in the travis builds.

Could you demonstrate that I'm wrong by making the change I requested, which will trigger another travis build?

Reply:
Have raised another PR with the changes and it works as you expected.